### ISSUE TITLE
HttpObjectDecoder.getChunkedSize trips over whitespace

### ISSUE THREAD

**Issue by lars-n** [netty/netty#13273]
### Expected behavior
the chunked size is calculated correctly

### Actual behavior
java.lang.NumberFormatException is thrown

### Steps to reproduce

See code below.

Please be aware that I'm not at all familiar with the netty code. Added a few comments in the code below.

### Minimal yet complete reproducer code (or URL to code)

```
import io.netty.util.internal.StringUtil;
import java.nio.charset.StandardCharsets;

public class JavaTester {

     private static int getChunkSizeNewFixed(byte[] hex, int start, int length) {
        // byte[] is produced by LineParse::parseLine that already skip ISO CTRL and Whitespace chars
        // COMMENT: ^ I could see handling for control chars but not whitespaces ^, but maybe I'm missing something.
        int result = 0;
        for (int i = 0; i < length; i++) {
            final int digit = StringUtil.decodeHexNibble(hex[start + i]);
            if (digit == -1) {
                // COMMENT: Until version 4.1.87.final the whitespace and control char checks where in there
                if (hex[start + i] == ';' || Character.isWhitespace(hex[start + i])) {
                    return result;
                }
                throw new NumberFormatException();
            }
            result *= 16;
            result += digit;
        }
        return result;
    }

    private static int getChunkSizeNew(byte[] hex, int start, int length) {
        // byte[] is produced by LineParse::parseLine that already skip ISO CTRL and Whitespace chars
        int result = 0;
        for (int i = 0; i < length; i++) {
            final int digit = StringUtil.decodeHexNibble(hex[start + i]);
            if (digit == -1) {
                // uncommon path
                if (hex[start + i] == ';') {
                    return result;
                }
                throw new NumberFormatException();
            }
            result *= 16;
            result += digit;
        }
        return result;
    }

    public static void main(String... args) {
        String toTest = "BDC Cookie: NSC_wTfswfs_c2cjoufhsbujpo=ffffffff099b5f1f45525d5f4f58455e445a4a42378b;expires=Mon, 13-Mar-2023 14:16:06 GMT;path=/;secure;httponly";
        System.out.println(getChunkSizeNewFixed(toTest.getBytes(StandardCharsets.UTF_8), 0, 4));
        System.out.println(getChunkSizeNew(toTest.getBytes(StandardCharsets.UTF_8), 0, 4)); <-- throws NumberFormatException
    }
}
```

### Netty version

from 4.1.88 onwards.

### JVM version (e.g. `java -version`)
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_292-b10)

### OS version (e.g. `uname -a`)
MacOS and Ubuntu

**Comment by normanmaurer** [comment]
@franz1981 PTAL... I think this might be related to a change of yours

**Comment by lars-n** [comment]
The `StringUtil.decodeHexNibble(hex[0 + 3])` call will be `StringUtil.decodeHexNibble(32)` which returns `-1`. Forgot to mention that above.

**Comment by franz1981** [comment]
mmm I don't follow it sorry...
@lars-n are you able to make this to be triggered by actual HTTP data?
The comment I've left for myself of the future (i.e. today)
```
byte[] is produced by LineParse::parseLine that already skip ISO CTRL and Whitespace chars
```
means that there's no need to check for whitespaces because are already checked upfront
i.e. https://github.com/netty/netty/blob/3de03282d558d7b7e2c282b31db73335f0f498b1/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java#L1123
that's been used by `HttpObjectDecoder::skipControlChars` and is populated by both white spaces and CTRL chars

**Comment by lars-n** [comment]
I'm beginning to wonder if the server I'm querying is returning crap data. Maybe pause for now while I'll try to verify that.

**Comment by lars-n** [comment]
@franz1981 , had a look at the code around `SKIP_CONTROL_CHARS_BYTES` you mentioned. The case where is breaks is for `32` which, correct me if I'm wrong, would not be skipped by that code? I think `32` is a whitespace in ASCII.

**Comment by franz1981** [comment]
Yeah, the part I don't understand is the skipping, tbh, that should have worked. Will take a look

**Comment by franz1981** [comment]
Thanks @lars-n for the hints, I've sent https://github.com/netty/netty/pull/13274, if you can take a look and try it would be great

**Comment by lars-n** [comment]
Is this the line that should be dropping the whitespace? https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java#L1085

In my case `skipControlChars` isn't called because `currentState` is set to `READ_CHUNK_SIZE` and so `currentState == State.SKIP_CONTROL_CHARS` evals to `false`. But again, really not familiar with the code here.

**Comment by franz1981** [comment]
Sadly it doesn't drop the white space but stop the very first char that's not a white space instead.
In the pr I have fixed the comment to better understand it next time.

**Comment by franz1981** [comment]
@lars-n let me know if this work for you, it is now working as the previous version, but faster

**Comment by lars-n** [comment]
Thanks for the quick response.

And a dumb question, is there a simple way for me to test this, except building netty myself?

**Comment by franz1981** [comment]
Mmmm probably not a fast one, sadly :(
I have added few tests covering your (and others) test cases but these are not end 2 end tests

**Comment by lars-n** [comment]
No worries, wasn't that difficult.

Can confirm that the change from the link PR fixes this issue.

**Comment by franz1981** [comment]
many thanks @lars-n for checking!

**Comment by lars-n** [comment]
you're welcome. Any idea on when this fix will be released? Simply so that I can plan here on whether we need to downgrade or can wait for the release.

**Comment by franz1981** [comment]
@normanmaurer ?

**Comment by normanmaurer** [comment]
Today or tomorrow

**Comment by lars-n** [comment]
yes, thanks a lot
